core/flyout_horizontal.js

```diff
@@ -236,44 +238,103 @@ Blockly.HorizontalFlyout.prototype.wheel_ = function(e) {
+      } else if (item instanceof Blockly.FlyoutButton) {
+        contentWidth += item.width;
+      }
+      contentWidth += gaps[i];
+    }
+    var delta = this.getWidth() - contentWidth;
+    if (delta > 0) {
+      cursorX += delta;
+    }
   }
-
+  
   for (var i = 0, item; (item = contents[i]); i++) {
-    if (item.type == 'block') {
-      var block = item.block;
-      var allBlocks = block.getDescendants(false);
-      for (var j = 0, child; (child = allBlocks[j]); j++) {
-        // Mark blocks as being inside a flyout.  This is used to detect and
-        // prevent the closure of the flyout if the user right-clicks on such a
-        // block.
-        child.isInFlyout = true;
+    if (item instanceof Blockly.BlockSvg) {
+      var block = item;
+      var dimensions = block.getHeightWidth();
+      var position = block.getRelativeToSurfaceXY();
+
+      var destinationX = cursorX;
+      if (block.outputConnection && !this.RTL) {
+        destinationX -= this.tabWidth_;
       }
-      block.render();
-      var root = block.getSvgRoot();
-      var blockHW = block.getHeightWidth();
 
-      // Figure out where to place the block.
-      var tab = block.outputConnection ? this.tabWidth_ : 0;
       if (this.RTL) {
-        var moveX = cursorX + blockHW.width;
-      } else {
-        var moveX = cursorX - tab;
+        destinationX += dimensions.width;
+      }
+      var destinationY = cursorY;
+
+      var deltaX = destinationX - position.x;
+      var deltaY = destinationY - position.y;
+
+      if (deltaX || deltaY) {
+        block.moveBy(deltaX, deltaY);
       }
-      block.moveBy(moveX, cursorY);
+      this.moveRectToBlock_(block.flyoutRect_, block);
+    } else if (item instanceof Blockly.FlyoutButton) {
+      item.moveTo(cursorX, cursorY);
+      var dimensions = item;
+    }
 
-      var rect = this.createRect_(block, moveX, cursorY, blockHW, i);
-      cursorX += (blockHW.width + gaps[i]);
+    cursorX += dimensions.width + gaps[i];
+  }
+  this.workspace_.setResizesEnabled(true);
+};
 
-      this.addBlockListeners_(root, block, rect);
-    } else if (item.type == 'button') {
-      this.initFlyoutButton_(item.button, cursorX, cursorY);
-      cursorX += (item.button.width + gaps[i]);
+/**
+ * Respond to changes in the dimensions of blocks by repositioning the contents
+ *     of the flyout.
+ * @param {!Blockly.Events.BlockChange} event The block change event.
+ * @protected
+ */
+Blockly.HorizontalFlyout.prototype.handleBlockChange_ = function(event) {
+  if (event.type === Blockly.Events.BLOCK_CHANGE) {
+    this.layoutContent_();
+    
+    // If we're RTL, rerender the text fields on this block (which were being
+    // edited to trigger this event) to fix their alignment relative to their
+    // parent block, which was just moved in the call to layoutContent_ above.
+    // This avoids the editing field falling out of alignment with the field
+    // rendered on the block itself.
+    if (this.RTL) {
```

Comment:
Why does this not apply for LTR as well?

I'm asking in the hopes that it does apply to both and we can just move this method to the parent class, since we do the same things in `VerticalFlyout`.